gitpod-io/gitpod
# [supervisor] Make sure we read the last chunk of message when task is closed

## Description
This PR fixes multiple issues:
 - [3b6ac9c](https://github.com/gitpod-io/gitpod/pull/20062/commits/3b6ac9c6ef558dcb24beea4829b4b568bd16ade1): removes (yet another) race in supervisor which might have lead to us not sending the last piece(s) of a log
 - [835ec3a](https://github.com/gitpod-io/gitpod/pull/20062/commits/835ec3a23ce963a0ede559cac954216745e77930):
   - forward the bytes array (`UInt8Array`) directly to xterm, instead of converting it to a `string` between: This could "chop" terminal control characters in between, leading to weird behavior like in ENT-528
   - ensure we close the response only after we completed pushing every chunk in the queue [[1](https://github.com/gitpod-io/gitpod/blob/e5003ecdd0ad04c26e50bc9df787e644546eccb1/components/server/src/workspace/headless-log-controller.ts#L207-L208) and [2](https://github.com/gitpod-io/gitpod/blob/e5003ecdd0ad04c26e50bc9df787e644546eccb1/components/server/src/workspace/headless-log-controller.ts#L369-L386)]
 - [f94a7f3](https://github.com/gitpod-io/gitpod/pull/20062/commits/f94a7f3df81426a4858cc4c189cb1692165d4551): don't call `res.end()` in server, even after sending the last chunk, but wait 30s before doing so.
   - not exactly sure what's going on, but I suspect because the connection browser -> caddy is the slowest, all these chunks end up being buffered in caddy
   - once we cancel the stream server-side, instead of forwarding all non-read chunks to the client, caddy cancels the downstream connection directly
     - I can't prove this, but [this issue](https://github.com/caddyserver/caddy/issues/4922) sounds very close
     - we could fit this with a plugin, but not here and now
 - [d70cf57](https://github.com/gitpod-io/gitpod/pull/20062/commits/d70cf573a7355cda83e4503504044b3e3bcc162c): when receiving a "headless log status code", we ignored potential prefix of the actual RegEx match, chopping of the last chunk of the result. This change extracts that and forward it to xterm before closing the stream


Future work:
 - debounce/buffer log chunks before sending to clients, instead of sending every piece that's written to the log file Ideally done in supervisor directly. Chunk size should be small enough to not overwhelm node.js, though (~1KB).
   - motivation: We have seen a delay between receiving an update in server to sending it to browser of 16s (!)
 - once started streaming, the frontend should only stop streaming if we navigate away from the page, not on status change (see [shouldFetchLogs](https://github.com/gitpod-io/gitpod/blob/e5003ecdd0ad04c26e50bc9df787e644546eccb1/components/dashboard/src/data/prebuilds/prebuild-logs-emitter.ts#L36-L107))

## Related Issue(s)

Fixes ENT-479
Fixes ENT-528

## How to test
 - use Firefox
 - start a prebuild on https://gpl-ft-fix11caf8d206.preview.gitpod-dev.com/workspaces based on this repo: https://github.com/filiptronicek/tests-with-spring-petclinic
 - open the view in 4 different browser windows
 - see how all logs successfully complete :heavy_check_mark: 

## Documentation


#### Preview status

gitpod:summary

## Build Options

<details>
<summary>Build</summary>

      Run the build with werft instead of GHA
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

</details>

<details>
<summary>Installer</summary>

  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

      If enabled this will build `install/preview`
      If enabled this will create the environment on GCE infra
      Saves cost. Untick this only if you're really sure you need a non-preemtible machine.
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
</details>

/hold
